edge-modules/api-proxy-module/src/monitors/config_parser.rs

```diff
@@ -9,90 +9,89 @@ use crate::token_service::token_server;
+        //reserved value cannot be changed by the user.
+        let mut reserved_values = HashMap::new();
 
         //Set default config values
         for (key, value) in PROXY_CONFIG_DEFAULT_VALUES.iter() {
             default_values.insert((*key).to_string(), (*value).to_string());
         }
 
-        let re = Regex::new(r"\$\{(.*?)\}").context("Failed to match variables")?;
-
-        for key in re.captures_iter(config) {
-            let new_key = key[1].to_string();
-
-            match std::env::var(key[1].to_string()) {
-                //If user passed the variable value through env var, select it
-                Ok(value) => {
-                    context.insert(new_key, (*value).to_string());
-                }
-                //Else check if it has a default value, otherwise set to 0
-                Err(_) => {
-                    if let Some(val) = default_values.get(&new_key) {
-                        context.insert(new_key, (*val).to_string());
-                    } else {
-                        context.insert(new_key, "0".to_string());
-                    }
-                }
-            };
-        }
-
         //Harded coded keys
-        if let Ok(moduleid) = env::var("IOTEDGE_PARENTAPIPROXYNAME") {
-            context.insert(
-                "IOTEDGE_PARENTAPIPROXYNAME".to_string(),
-                sanitize_dns_label(&moduleid),
-            );
-        }
-
         // Tokens are cached by nginx. Current duration is 30mn. Make sure the the validity is not lower than the caching time.
-        context.insert(
+        reserved_values.insert(
             "TOKEN_VALIDITY_MINUTES".to_string(),
             (token_server::TOKEN_VALIDITY_SECONDS / 60 / 2).to_string(),
         );
-        context.insert(
+        reserved_values.insert(
             "TOKEN_SERVER_PORT".to_string(),
             token_server::TOKEN_SERVER_PORT.to_string(),
         );
 
-        Ok(ConfigParser { context })
+        Ok(ConfigParser {
+            default_values,
+            reserved_values,
+        })
     }
 
-    pub fn get_key_value(&self, key: &str) -> Option<(String, String)> {
-        if let Some(key_pair) = self.context.get_key_value(key) {
-            return Some((key_pair.0.to_string(), key_pair.1.to_string()));
-        }
+    pub fn load_config(&self, config: &str) -> Result<HashMap<String, String>, anyhow::Error> {
+        let mut context = HashMap::new();
 
-        None
-    }
+        let re = Regex::new(r"\$\{(.*?)\}").context("Failed to match variables")?;
 
-    pub fn add_new_key(&mut self, key: &str, val: &str) {
-        self.context.insert(key.to_string(), val.to_string());
-    }
+        for key in re.captures_iter(config) {
+            let new_key = key[1].to_string();
 
-    pub fn remove_key(&mut self, key: &str) {
-        self.context.remove(key);
-    }
+            // This priority could be coded as a nested if else statement but it is hard to read.
```

Comment:
Makes sense now. Thanks!
I think if you invert your steps you may be more concise code like this
```
            let new_key = key[1].to_string();
            let new_value = self
                .reserved_values
                .get(&new_key)
                .map(|v| v.to_string())
                .or_else(|| std::env::var(&new_key).ok())
                .or_else(|| self.default_values.get(&new_key).map(|v| v.to_string()))
                .unwrap_or(String::from("0"));

            context.insert(new_key, new_value);
```

Reply:
Looks great thanks! I am not used to declarative style yet.